[VL] Error: Native split: shuffle writer stop failed - Merging from spilled file NO.7 is not exhausted

### Backend

VL (Velox)

### Bug description

reproduce using gluten-it

```
<gluten-it-bin> queries --local --preset=velox --benchmark-type=ds --error-on-memleak --disable-aqe --off-heap-size=5g -s=100.0 --shuffle-partitions=72 --threads=8 --iterations=1 --queries=q67 --enable-history --enable-ui
```

### Spark version

None

### Spark configurations

3.2.2

### System information

_No response_

### Relevant logs

```bash
I0831 15:19:27.594770 497885 VeloxBackend.cc:43] session conf = {
 {spark.gluten.ugi.username = root}
 {spark.sql.caseSensitive = false}
 {spark.gluten.sql.session.timeZone.default = Asia/Shanghai}
 {spark.gluten.memory.task.offHeap.size.in.bytes = 671088640}
 {spark.gluten.ugi.tokens = }
}
23/08/31 15:19:27 ERROR TaskResources: Task 100 failed by error: 
java.lang.RuntimeException: Native split: shuffle writer stop failed - Merging from spilled file NO.7 is not exhausted.
	at io.glutenproject.vectorized.ShuffleWriterJniWrapper.stop(Native Method)
	at org.apache.spark.shuffle.ColumnarShuffleWriter.internalWrite(ColumnarShuffleWriter.scala:177)
	at org.apache.spark.shuffle.ColumnarShuffleWriter.write(ColumnarShuffleWriter.scala:218)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1491)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
23/08/31 15:19:27 ERROR Executor: Exception in task 69.0 in stage 27.0 (TID 100)
java.lang.RuntimeException: Native split: shuffle writer stop failed - Merging from spilled file NO.7 is not exhausted.
	at io.glutenproject.vectorized.ShuffleWriterJniWrapper.stop(Native Method)
	at org.apache.spark.shuffle.ColumnarShuffleWriter.internalWrite(ColumnarShuffleWriter.scala:177)
	at org.apache.spark.shuffle.ColumnarShuffleWriter.write(ColumnarShuffleWriter.scala:218)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1491)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
```
```

Another kind of error but same cause

```
Caused by: java.lang.RuntimeException: Native split: shuffle writer stop failed - Read out of bounds (offset = 139932668077264, size = 139937618202384) in file of size 186800
        at io.glutenproject.vectorized.ShuffleWriterJniWrapper.stop(Native Method)
        at org.apache.spark.shuffle.ColumnarShuffleWriter.internalWrite(ColumnarShuffleWriter.scala:177)
        at org.apache.spark.shuffle.ColumnarShuffleWriter.write(ColumnarShuffleWriter.scala:218)
        at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
        at org.apache.spark.scheduler.Task.run(Task.scala:131)
        at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1491)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:750)
```

--- a/cpp/core/shuffle/LocalPartitionWriter.cc
+++ b/cpp/core/shuffle/LocalPartitionWriter.cc
@@ -312,14 +312,16 @@ arrow::Status PreferCachePartitionWriter::stop() {
     ARROW_ASSIGN_OR_RAISE(auto startInFinalFile, dataFileOs_->Tell());
     // 4. Iterator over all spilled files
     for (auto i = 0; i < spills_.size(); ++i) {
-      auto partitionSpillInfo = spills_[i].partitionSpillInfos[spillInfoOffsets[i]];
-      // 5. read if partition exists in the spilled file and write to the final file
-      if (partitionSpillInfo.partitionId == pid) { // A hit
-        firstWrite = false;
-        ARROW_ASSIGN_OR_RAISE(auto raw, spilledFiles[i]->ReadAt(partitionSpillInfo.start, partitionSpillInfo.length));
-        RETURN_NOT_OK(dataFileOs_->Write(raw));
-        // Goto next partition in this spillInfo
-        spillInfoOffsets[i]++;
+      if (spillInfoOffsets[i] < spills_[i].partitionSpillInfos.size()) {
+        auto partitionSpillInfo = spills_[i].partitionSpillInfos[spillInfoOffsets[i]];
+        // 5. read if partition exists in the spilled file and write to the final file
+        if (partitionSpillInfo.partitionId == pid) { // A hit
+          firstWrite = false;
+          ARROW_ASSIGN_OR_RAISE(auto raw, spilledFiles[i]->ReadAt(partitionSpillInfo.start, partitionSpillInfo.length));
+          RETURN_NOT_OK(dataFileOs_->Write(raw));
+          // Goto next partition in this spillInfo
+          spillInfoOffsets[i]++;
+        }
       }
     }
     // 6. Write cached batches

